### ISSUE TITLE
Latest update broke the app

### ISSUE THREAD

**Issue by ux-engineer** [jasonlewis/basset#159]
...fixing it. Just split the configs.

Now getting:

Argument 2 passed to Basset\Environment::make() must be an instance of Closure, array given, called in /volume1/web/appname/vendor/jasonlewis/basset/src/Basset/Environment.php on line 163 and defined 

Update: Needed to remove one array wrapping, so it's working again.

Could it be possible to add a kind of "you need to do this after you did this composer update" note when updating via Composer?

**Comment by ivannpaz** [comment]
The previous config was something along the lines of:

``` php
return array(
    'collections' => array(
        'application' => function ($collection)
        {
            //configuration
        }
    ),
    //...
    'production'    => array('production'),
    //...
    'aliases' => array(
        'assets' => array(...),
        'filters' => array(...),
    ),
);
```

Now just make sure the new config.php does not contain the `collections` or `aliases` keys (they are mainly ignored afaik)

Create the new two config files:

``` php
//collections.php
return array(
    'application' => function ($collection)
    {
        //configuration
    },
);
```

And:

``` php
//aliases.php
return array(
    'assets' => array(...),
    'filters' => array(...),
);
```

**Comment by ux-engineer** [comment]
It seems alright. But building via Artisan gives:
Starting development build...
[collection] Collection not found.

Now that I deployed in productions I can see also that the files are not combined. Environment settings are alright.

Update: Two collections have duplicate items, dropping the other one now...

**Comment by jasonlewis** [comment]
I actually made it build development by default and switched it to a
production flag.

Try php artisan basset:build --production

**Comment by jasonlewis** [comment]
I wish I could alert people of changes and try not to do too many breaking changes. If you feel any of these changes weren't good choices please let me know your thoughts. You guys are using it as well. Knowing how you use it is crucial to me.

**Comment by ux-engineer** [comment]
I still haven't figured it out why this production is broken...

Update: Now something happened... Lets see :)

**Comment by jasonlewis** [comment]
Does using the --production switch not work?

**Comment by ux-engineer** [comment]
Now I got the missing files, might have been those duplicates. But the files aren't still combined - was there some special trick for that? :)

I did try with swapping the environment setting in dev environment, not combining.

**Comment by ux-engineer** [comment]
When I deleted all the built files in production and tried basset:build, it says "build was not required for collection." ?

Update: I needed to delete /storage/meta/collections.json for rebuilding. But then: file_put_contents([....]/app/storage/meta/collections.json): failed to open stream: Permission denied

**Comment by jasonlewis** [comment]
Make sure you have writeable permissions of said directory so Basset can write the collections file.

**Comment by jasonlewis** [comment]
On my copy building a production collection and switching my environment to `production` correctly combines and displays the combined production collection in my view.

**Comment by ux-engineer** [comment]
I figured out the permissions bit. Don't know how to debug why this won't combine and minimize on production...

**Comment by jasonlewis** [comment]
Are you performing the build on your production environment or somewhere else, then pushing to production?

Can you also [paste your aliases](http://paste.laravel.com).

**Comment by ux-engineer** [comment]
Built on development environment and uploaded to server.

Aliases here http://paste.laravel.com/vU7

**Comment by jasonlewis** [comment]
Hm, this is odd.

Is the build succeeding? Or you don't get the built file at all? Paste your collections as well. I'll try to recreate this.

**Comment by ux-engineer** [comment]
I get the build files, but shouldn't they be combined to one rather than singular files like on dev environment. Some days ago I tried swapping the environments and the files got combined then, but not anymore. Weird...

Here are the collections: http://paste.laravel.com/vUr

**Comment by jasonlewis** [comment]
This might not be the reason but take a look at this updated file: http://paste.laravel.com/vUt

When you call `$collection->directory()` you get an instance of `Basset\Directory` and as such you shouldn't really overwrite the `$collection` variable. It also seemed like you were trying to apply filters to specific directories but applying it to the entire collection. Take a look at the changes to the file. I've left a couple of comments here and there as well.

---

The build files _should_ be combined in to one file when running a _production_ build. If their not, then it sounds like you're not running a production build. Things to check.
1. Update Basset again (check your `composer.json`, at the moment it might be easier if you use `dev-master` as the version).
2. Check your `bootstrap/start.php` and the environment settings. You should have a `production` environment that will match the production environment listed in the Basset config.
3. Run `php artisan basset --delete-manifest` to delete the manifest file and then try to build again.
4. Turn on `debug` in the Basset config. You may need to delete the manifest file and try to rebuild again.
5. Change your local environment to the production environment so you can test that the combined file is being included in your views.

Remember you should be running `php artisan basset:build collection --production`

**Comment by ux-engineer** [comment]
Did all that and tried with chowning 777 recursively to builds folder :/

**Comment by jallits** [comment]
@jasonlewis I have no preference on what the collection default builds are. However I would request to have both develop (-d) and a prod (-p) flags when I build my collections from Artisan. This way I get the warm fuzzies of knowing what I am actually building and I feel like a script kiddie because I can use a flag in the command line. (J/K)

**Comment by jallits** [comment]
This is what I had to do to get the latest.

---

BACKUP YOUR OLD CONFIG FILE SOMEWHERE ON YOUR FILESYSTEM
FROM /app/packages/jasonlewis/basset

---

Manually delete the /public/*builds directory.
Manually delete the collections.json file in /app/storage/meta
Manually delete the basset directory in /app/packages/jasonlewis

RUN `php artisan config:publish jasonlewis/basset`

Updated the collections.php file with the collection from your backed up config.php file.

**Comment by ux-engineer** [comment]
@jallits have you also had problems with not getting combined files on production environment?

**Comment by jasonlewis** [comment]
I'm not really sure what your problem is. Double then triple check your environments and filters. Try to build it locally. It's working fine here. I've even [deployed my site](http://jasonlewis.me) which uses Basset and it builds perfectly.

**Comment by jallits** [comment]
@envisio Nope everything worked as advertised with the almost-infamous "Split Config" commit. I just posted the steps that I took to get things in order in the hopes that it might get you back on track.

Of course, when I woke up this morning I saw @jasonlewis tweet that makes this all kind of mute.

**Comment by ivannpaz** [comment]
Sidenote: I should also add that after the split, for me it was working perfectly ok... even the combining on production deploy. If possible, it would be great if @jasonlewis creates a development branch for all the new changes, or at least a "mid stable" tag that we can lock on composer until further notice?

**Comment by ux-engineer** [comment]
@jallits @jasonlewis Actually this not-combining problem was before the update, sorry that I forgot to mention about it.

I am re-building the app on a fresh Laravel install. Just noting at this stage that Basset is now giving error that the files are not in "assets" folder. Where they weren't earlier but the collections were working fine then, only not combining on production.

Update: The weirdest part is that (after moving the files to default locations) building from artisan builds only one collection named 'application'. Which is a collection I had earlier but removed. The other 7 collections are not found even they are on collection file. That 'application' isn't even in the collections config.

Now I'm getting logged error for filters: [2013-06-11 22:20:28] basset.ERROR: Failed to find required constructor argument for filter "Sass\ScssFilter". (Parameter #0 [ <optional> $sassPath = '/usr/bin/sass' ])

Also [2013-06-11 22:20:28] basset.ERROR: Directory "coffeescripts" could not be found in "/volume1/web/projectname/public/assets/javascripts" [] []   ...I haven't used those filters and I commented out their declarations from aliases.php

Update: artisan config:publish jasonlewis/basset seems to publish the old style configs. Composer update was with dev-master after the fresh install.

**Comment by jallits** [comment]
@envisio Sorry I wasn't clearer about this in my last comment about being a mute point. Yes, @jasonlewis reverted from split config files to a single config file in his last commit. He tweeted it earlier, and apologized for any difficulties on our end. That's the price we pay for being on dev-master.

Can you do a pastebin for the following or email me: dan [at] jallits [dot] com

/app/config/packages/jasonlewis/basset/config.php
/app/config/app.php
/composer.json

**Comment by ux-engineer** [comment]
I emailed those files to @jallits. Meanwhile updating here that I did a fresh install again. Then added only Basset (as dev-master) and ExpressiveDate packages via Composer update. Added those to service providers array and Basset as facade in app.php config file. Then ran 'artisan config:publish jasonlewis/basset' and I'm getting singular config file for Basset? :facepunch:

**Comment by jallits** [comment]
Yes as I mentioned previously Jason reverted back to using only 1 config file. This is correct. Send me your new config.php file. Also could you share with me the directory structure of your public directory.

**Comment by ux-engineer** [comment]
Oh sorry, seems I had missed that while I was elsewhere for those few days. It's all working now - the issue was the --production flag on the command for building collections. I had expected that building collections also built files for production as default without needing a flagged command.

Thanks again for your prompt support! :)

**Comment by jallits** [comment]
Awesome. Glad it works for you now!

**Comment by ux-engineer** [comment]
...yet I don't see minification happening. This is production settings on now, and I'm seeing combined files. @jasonlewis already checked my configs here which we shared here.

Are all the libraries needed for minification installed via Composer? I'm using only minify on CSS and JS.

**Comment by jallits** [comment]
How did you install your minification libraries? Some people use Node's NPM or you can choose PHP ports. Check out Jason's composer.json file for how he includes [CSSmin](https://github.com/jasonlewis/website/blob/master/composer.json)

**Comment by ux-engineer** [comment]
The thing is, there isn't any mentioning in the doc about needing to install custom minification libraries. One would assume such dependencies are handled by composer when installing Basset. @jasonlewis ?

Installed "mrclay/minify": "dev-master" which does minify the JS files. This library should work also for CSS but in this case it's not. Seems that this dependency has been removed from Assetic, which seems to use uglify libraries: https://github.com/kriswallsmith/assetic/blob/master/package.json Why is Assetic needed anyways?

When I install CSSmin like Jason has and build production collections, I'm losing most of the styles from the site. This weird as I haven't seen this kind of problem before. The minified CSS seems fine. I've had media queries break styles when minified or concat but I checked those, and in that case I should still get more styles visible. Right now there is only base Bootstrap styles visible on the form elements. Perhaps Bootstrap is using some too exotic markup for CSSmin to handle... :palm_tree: 

Installed also natxet/CssMin instead of CSSmin. It is the most popular CSSmin package from Packagist. It did nothing.
